File: control-plane/ap-controller/app/audit.py
import json
import time
import hmac
import hashlib
from typing import Dict, Any

from app.config import load_config

_cfg = load_config()
_audit_cfg = _cfg.get("audit", {})

AUDIT_ENABLED = _audit_cfg.get("enabled", True)
AUDIT_SECRET = _audit_cfg.get("secret")

if not AUDIT_SECRET:
    raise RuntimeError("audit.secret is missing in controller.yaml")


def sign_audit(payload: Dict[str, Any]) -> str:
    raw = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hmac.new(
        AUDIT_SECRET.encode(),
        raw.encode(),
        hashlib.sha256
    ).hexdigest()


def audit_log(event: str, **context):
    if not AUDIT_ENABLED:
        return

    record = {
        "ts": int(time.time()),
        "event": event,
        **context,
    }

    record["sig"] = sign_audit(record)

    # stdout → docker logs → log pipeline
    print(json.dumps(record, separators=(",", ":")))

File: control-plane/ap-controller-go/internal/audit/audit.go
package audit

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"time"
)

func New(enabled bool, secret string) *Logger {
	return &Logger{
		Enabled: enabled,
		Secret:  []byte(secret),
	}
}

func (l *Logger) Sign(payload []byte) string {
	m := hmac.New(sha256.New, l.Secret)
	m.Write(payload)
	return hex.EncodeToString(m.Sum(nil))
}

func (l *Logger) Write(event map[string]any) {
	if !l.Enabled {
		return
	}
	if _, ok := event["ts"]; !ok {
		event["ts"] = time.Now().Unix()
	}
	// 先不带 sig 序列化
	tmp := make(map[string]any, len(event))
	for k, v := range event {
		if k == "sig" {
			continue
		}
		tmp[k] = v
	}
	b, _ := json.Marshal(tmp)
	sig := l.Sign(b)

	tmp["sig"] = sig
	out, _ := json.Marshal(tmp)
	// 写 stdout，docker logs 里就是 JSON（你现在就是这么看的）
	_, _ = os.Stdout.Write(append(out, '\n'))
}
